### PR TITLE
Add Translation Natives

### PR DESCRIPTION
**Pull request by Headline**
See #626 for the discussion.

### CODE REVIEW

**Comment by WildCard65** [comment]
Any changes to SM public interfaces REQUIRE a version macro value increase
```cpp
#define SMINTERFACE_TRANSLATOR_VERSION	4
```
must become
```cp
#define SMINTERFACE_TRANSLATOR_VERSION	5
```

**Comment by Headline22** [comment]
Thanks for your comment, I had a feeling there was something I was missing in regards to the Translator interface. I'll commit the change in a little while

File: core/logic/PhraseCollection.cpp [hunk 1]
```diff
@@ -89,6 +89,21 @@ IPhraseFile *CPhraseCollection::GetFile(unsigned int file)
 	return m_Files[file];
 }
 
+bool CPhraseCollection::TranslationPhraseExists(const char *key)
+{
+	size_t i;
+
+	for (i = 0; i < m_Files.size(); i++)
```
Comment (KyleSanderson) [inline core/logic/PhraseCollection.cpp]: size_t

File: core/logic/smn_lang.cpp [hunk 2]
```diff
@@ -36,6 +36,37 @@
 #include <ISourceMod.h>
 #include <am-string.h>
 
+static cell_t sm_TranslationPhraseExists(IPluginContext *pCtx, const cell_t *params)
+{
+	IPlugin *pl = pluginsys->FindPluginByContext(pCtx->GetContext());
+	IPhraseCollection* collection = pl->GetPhrases();
```
Comment (KyleSanderson) [inline core/logic/smn_lang.cpp]: Weird type style.

File: core/logic/smn_lang.cpp [hunk 3]
```diff
@@ -36,6 +36,37 @@
 #include <ISourceMod.h>
 #include <am-string.h>
 
+static cell_t sm_TranslationPhraseExists(IPluginContext *pCtx, const cell_t *params)
+{
+	IPlugin *pl = pluginsys->FindPluginByContext(pCtx->GetContext());
+	IPhraseCollection* collection = pl->GetPhrases();
+	
+	char *phrase;
+
```
Comment (KyleSanderson) [inline core/logic/smn_lang.cpp]: Crunch.

File: core/logic/smn_lang.cpp [hunk 4]
```diff
@@ -36,6 +36,37 @@
...
+	IPlugin *pl = pluginsys->FindPluginByContext(pCtx->GetContext());
+	IPhraseCollection* collection = pl->GetPhrases();
+	
+	char *phrase;
+	pCtx->LocalToString(params[1], &phrase);
+	
+	int langid = params[2];
+
+	Translation trans;
+	if (collection->FindTranslation(phrase, langid, &trans) == Trans_Okay)
```
Comment (KyleSanderson) [inline core/logic/smn_lang.cpp]: return (collection->FindTranslation(phrase, langid, &trans) == Trans_Okay);

File: public/ITranslator.h [hunk 5]
```diff
@@ -221,6 +230,16 @@ namespace SourceMod
 			unsigned int numparams,
 			size_t *pOutLength,
 			const char **pFailPhrase) =0;
+
```
Comment (KyleSanderson) [inline public/ITranslator.h]: Extra newline (?).

File: core/logic/Translator.cpp [hunk 6]
```diff
@@ -672,6 +672,15 @@ TransError CPhraseFile::GetTranslation(const char *szPhrase, unsigned int lang_i
 	return Trans_Okay;
 }
 
+bool CPhraseFile::TranslationPhraseExists(const char *phrase)
+{
+	int address;
+	if (!m_PhraseLookup.retrieve(phrase, &address))
```
Comment (KyleSanderson) [inline core/logic/Translator.cpp]: return m_PhraseLookup.retrieve(phrase, &address);

File: public/ITranslator.h (hunk 5 above)
Comment (Headline) [reply public/ITranslator.h]: GitHub thinks this \n wasn't removed, but it was. Thanks for your comments

File: public/ITranslator.h [hunk 7]
```diff
@@ -113,6 +113,15 @@ namespace SourceMod
 		 * @return					File name.
 		 */
 		virtual const char *GetFilename() =0;
+		
+		/**
+		 * @brief Determines if a translation phrase exists within
+		 * the file.
+		 *
+		 * @param phrase				Phrase to search for.
+		 * @return						True if search was successful.
```
Comment (KyleSanderson) [inline public/ITranslator.h]: Last one, this doesn't read quite right to me and doesn't match the sp definition. I think the sp docs might be a little more consistent (s/search/phrase/g imo).

**Comment by joao7yt** [comment]
do anyone know when this will be added to the stable build?

**Comment by Benoist3012** [comment]
when SM 1.9 is ready to launch on stable branch, for now it's on dev branch only. And I don't think any backport is planned so you have to wait til sm 1.9 lands on stable branch.

**Comment by joao7yt** [comment]
just so i can understand: why some dev builds are making to stable builds? i can see that build 6220 (dev) was included on build 6038 (stable)

**Comment by Benoist3012** [comment]
You seem to misunderstand build & changes. These two builds aren't the same, they only share the same change, if you also look carefully the changes have been made on two different branches, you can consider that as a backport. https://imgur.com/a/BheoM
And as I said earlier, no backport are planned for the changes from that PR. Only small stuff is backported, and these changes aren't small, so wait til sm 1.9 lands on stable branch.

**Comment by joao7yt** [comment]
Oh, now i get it, tks for your attention :)

**Comment by Benoist3012** [comment]
Glad I could help!

**Comment by Headline** [comment]
We only move push changes to stable branches if it’s a security, gamedata, or crash fix. This is why that update landed on 1.8 and the translation native didn’t. You can still use the dev branch on your servers and use the translations native, but you just have to keep in mind you’re using the current development branch and instability is possible.

1.9 will likely become stable in the near future anyway.

**Pull request merged**